File: LeetCode/cpp/18.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int> &nums, int target) {
        int size = nums.size();
        if (size == 0)
            return {};
        set<vector<int>> vec_set;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < size - 3; ++i) {
            for (int j = i + 1; j < size - 2; ++j) {
                int k = j + 1, l = size - 1, sum = nums[i] + nums[j];
                while (k < l) {
                    if (sum + nums[k] + nums[l] > target)
                        --l;
                    else if (sum + nums[k] + nums[l] < target)
                        ++k;
                    else {
                        vec_set.insert({nums[i], nums[j], nums[k], nums[l]});
                        --l;
                        ++k;
                    }
                }
            }
        }
        vector<vector<int>> ret;
        for (auto &s:vec_set)
            ret.push_back(s);
        return ret;
    }
};

File: LeetCode/python3/18.py
class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        size = len(nums)
        if size < 4:
            return []
        nums.sort()
        ret = []
        nums_set = set()
        for i in range(size - 3):
            for j in range(i + 1, size - 2):
                k, l = j + 1, size - 1
                while k < l:
                    sum = nums[i] + nums[j] + nums[k] + nums[l]
                    if sum > target:
                        l -= 1
                    elif sum < target:
                        k += 1
                    else:
                        if [nums[i], nums[j], nums[k], nums[l]] not in ret:
                            ret.append([nums[i], nums[j], nums[k], nums[l]])
                        k += 1
                        l -= 1
        return ret
